ToolBox/Benzinga/BenzingaNewsDataDownloader.cs

```diff
@@ -0,0 +1,235 @@
+    /// Downloads Benzinga news data from Benzinga's API
+    /// </summary>
+    public class BenzingaNewsDataDownloader
+    {
+        private const string _baseUrl = "https://api.benzinga.com/api/v2";
+        private const int _retryMaxCount = 5;
+
+        private readonly IMapFileProvider _mapFileProvider;
+        private readonly MapFileResolver _mapFileResolver;
+        private readonly DirectoryInfo _destinationDirectory;
+        private readonly RateGate _rateGate;
+        private readonly string _apiKey;
+
+        /// <summary>
+        /// Creates an instance of the class
+        /// </summary>
+        /// <param name="destinationDirectory">
+        /// Directory to write files to. This will be a top-level folder that will have folders
+        /// following the pattern: {destinationDirectory}/YYYY/MM/dd/entry.json
+        ///
+        /// This should normally be `/raw_data_folder/alternative/benzinga`
+        /// </param>
+        /// <param name="apiKey">Key to access Benzinga's API</param>
+        public BenzingaNewsDataDownloader(DirectoryInfo destinationDirectory, string apiKey)
+        {
+            _mapFileProvider = Composer.Instance.GetExportedValueByTypeName<IMapFileProvider>(Config.Get("map-file-provider", "LocalDiskMapFileProvider"));
+            _mapFileResolver = _mapFileProvider.Get(Market.USA);
+
+            _destinationDirectory = destinationDirectory;
+            _apiKey = apiKey;
+
+            // Limit ourselves to 2 requests per second to Benzinga's API.
+            // There are 237391 articles starting from 2017-09-12 and ending on 2019-09-13.
+            // This means that at 200 articles/second, it would take us approximately
+            // 1186.955 seconds (19 minutes 45 seconds) + some latency to download.
+            _rateGate = new RateGate(1, TimeSpan.FromSeconds(0.5));
+        }
+
+        /// <summary>
+        /// Downloads news data from Benzinga's API
+        /// </summary>
+        /// <param name="startDate">Starting date</param>
+        /// <param name="endDate">Ending date</param>
+        public void Download(DateTime startDate, DateTime endDate, bool forceOverwrite = true)
+        {
+            foreach (var date in Time.EachDay(startDate, endDate))
+            {
+                var articles = new List<BenzingaNews>();
+                var finished = false;
+                var page = 0;
+
+                var finalDateDirectory = new DirectoryInfo(
+                    Path.Combine(
+                        _destinationDirectory.FullName,
+                        date.ToStringInvariant("yyyyMMdd")
+                    )
+                );
+
+                finalDateDirectory.Create();
+
+                while (!finished)
+                {
+                    for (var retries = 0; retries < _retryMaxCount; retries++)
+                    {
+                        try
+                        {
+                            using (var client = new WebClient())
```

Comment:
Might be a good idea to create the web client once and re use it

Reply:
Just leaving a note here for any other readers: In the past, my attempts to re-use the `WebClient` has been met with hostility from the WebClient class with the error: "WebClient does not support concurrent I/O operations" in a synchronous environment. With that reason, I believe for stability reasons that it is a better idea to pay the penalty for creating a new `WebClient` per request rather than add an additional point of failure in the downloader code.

Knowing this, if you'd like for me to re-use the WebClient after I address this round of reviews, please let me know and I'll make the change 👍